sound/soc/sof/sof-audio.c

```diff
@@ -107,6 +107,17 @@ int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 	if (!swidget->private)
 		return 0;
 
+	/*
+	 * PCM streams can be freed without being restarted after system resume. Prevent
+	 * the use_count from being decremented below 0 for dynamic widgets and below 1
+	 * for static widgets in such scenarios.
+	 */
+	if (swidget->dynamic_pipeline_widget && !swidget->use_count)
+			return 0;
+
+	if (!swidget->dynamic_pipeline_widget && swidget->use_count == 1)
+			return 0;
```

Comment:
I understand the motivation, but this makes me a bit suspicious: use-counts are usually used to trace users and uses and sometimes to verify consistency, but less often as an indication of the system state. They are more "mostly-write" than "mostly-read" if it makes sense. In a way they should stay consistent because the flow is correct, not the flow is corrected to keep them consistent. So it would be better (if possible) here not to use `use_count` for decision making. Rather some other indication of state.

Reply:
@lyakh I gave this a lot of thought but unfortunately the PCM state is the same during hw_free() irrespective of whether it is called after a normal stream stop or after resuming from suspend. The FE hw_free() checks the spcm->prepared[dir] flag during hw_free() to avoid this problem. Its the BE hw_free() that's problematic. Maybe what I can do is introduce a prepapred flag in snd_sof_dai that can be used to check if hw_free() should do anything or not. stay tuned!